tests/cmake.py

```diff
@@ -81,15 +96,73 @@ def cmake(cwd, targets, options=None):
     except subprocess.CalledProcessError:
         pytest.fail("cmake configure failed")
 
+    # CodeChecker invocations and options are documented here:
+    # https://github.com/Ericsson/codechecker/blob/master/docs/analyzer/user_guide.md
+
     buildcmd = [*cmake, "--build", ".", "--parallel"]
     for target in targets:
         buildcmd.extend(["--target", target])
+    if "code-checker" in os.environ.get("RUN_ANALYZER", ""):
+        buildcmd = [
+            "CodeChecker",
+            "log",
+            "--output",
+            "compilation.json",
+            "--build",
+            " ".join(buildcmd),
+        ]
+
     print("{} > {}".format(cwd, " ".join(buildcmd)), flush=True)
     try:
         subprocess.run(buildcmd, cwd=cwd, check=True)
     except subprocess.CalledProcessError:
         pytest.fail("cmake build failed")
 
+    if "code-checker" in os.environ.get("RUN_ANALYZER", ""):
+        # For whatever reason, the compilation summary contains duplicate entries,
+        # one with the correct absolute path, and the other one just with the basename,
+        # which would fail.
+        with open(os.path.join(cwd, "compilation.json")) as f:
+            compilation = json.load(f)
+            compilation = list(filter(lambda c: c["file"].startswith("/"), compilation))
+        with open(os.path.join(cwd, "compilation.json"), "w") as f:
+            json.dump(compilation, f)
+
+        disable = [
+            "readability-magic-numbers",
+            "cppcoreguidelines-avoid-magic-numbers",
+            "readability-else-after-return",
+        ]
+        disables = []
+        for d in disable:
+            disables.append("--disable")
+            disables.append(d)
+        checkcmd = [
+            "CodeChecker",
+            "check",
+            "--jobs",
+            str(os.cpu_count()),
+            # NOTE: The clang version on CI does not support CTU :-(
+            # Also, when testing locally, CTU spews a ton of (possibly) false positives
+            # "--ctu-all",
+            # TODO: we currently get >300 reports with `enable-all`
+            # "--enable-all",
+            *disables,
+            "--print-steps",
+            "--ignore",
+            os.path.join(source_dir, ".codechecker-ignore"),
+            "--logfile",
+            "compilation.json",
+        ]
+        print("{} > {}".format(cwd, " ".join(checkcmd)), flush=True)
+        child = subprocess.run(checkcmd, stdout=subprocess.PIPE, cwd=cwd, check=True)
+        sys.stdout.buffer.write(child.stdout)
+        marker = b"Total number of reports: "
+        errors = child.stdout[child.stdout.rfind(marker) + len(marker) :]
+        errors = int(errors[: errors.find(b"\n")])
+        if errors > 0:
+            pytest.fail("code-checker analysis failed")
+
```

Comment:
minor: personally this function is now too big and long for me to remember all variables in scope etc and i would break this up.  at this point it's not problematic though, so i don't mind.